src/invariant/ConservationOfLumens.cpp

```diff
@@ -0,0 +1,123 @@
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "invariant/ConservationOfLumens.h"
+#include "database/AccountQueries.h"
+#include "invariant/InvariantManager.h"
+#include "ledger/LedgerDelta.h"
+#include "lib/util/format.h"
+#include "main/Application.h"
+#include <numeric>
+
+namespace stellar
+{
+
+std::shared_ptr<Invariant>
+ConservationOfLumens::registerInvariant(Application& app)
+{
+    return app.getInvariantManager().registerInvariant<ConservationOfLumens>();
+}
+
+std::string
+ConservationOfLumens::getName() const
+{
+    return "ConservationOfLumens";
+}
+
+bool
+ConservationOfLumens::isStrict() const
+{
+    return false;
+}
+
+int64_t
+ConservationOfLumens::calculateDeltaBalance(LedgerEntry const* current,
+                                            LedgerEntry const* previous) const
+{
+    assert(current || previous);
+    auto let = current ? current->data.type() : previous->data.type();
+    if (let == ACCOUNT)
+    {
+        return (current ? current->data.account().balance : 0) -
+               (previous ? previous->data.account().balance : 0);
+    }
+    return 0;
+}
+
+std::string
+ConservationOfLumens::checkOnOperationApply(Operation const& operation,
+                                            OperationResult const& result,
+                                            LedgerDelta const& delta)
+{
+    auto const& lhCurr = delta.getHeader();
+    auto const& lhPrev = delta.getPreviousHeader();
+
+    int64_t deltaTotalCoins = lhCurr.totalCoins - lhPrev.totalCoins;
+    int64_t deltaFeePool = lhCurr.feePool - lhPrev.feePool;
+    int64_t deltaBalances = std::accumulate(
+            delta.addedBegin(), delta.addedEnd(), static_cast<int64_t>(0),
+            [this] (int64_t lhs, LedgerDelta::AddedLedgerEntry const& rhs) {
+                return lhs + calculateDeltaBalance(&rhs.current->mEntry,
+                                                   nullptr);
+            });
+    deltaBalances += std::accumulate(
+            delta.modifiedBegin(), delta.modifiedEnd(), static_cast<int64_t>(0),
+            [this] (int64_t lhs, LedgerDelta::ModifiedLedgerEntry const& rhs) {
+                return lhs + calculateDeltaBalance(&rhs.current->mEntry,
+                                                   &rhs.previous->mEntry);
+            });
+    deltaBalances += std::accumulate(
+            delta.deletedBegin(), delta.deletedEnd(), static_cast<int64_t>(0),
+            [this] (int64_t lhs, LedgerDelta::DeletedLedgerEntry const& rhs) {
+                return lhs + calculateDeltaBalance(nullptr,
+                                                   &rhs.previous->mEntry);
+            });
+
+    if (result.tr().type() == INFLATION)
+    {
+        int64_t inflationAmount = std::accumulate(
```

Comment:
This name is misleading - inflation amount is sum of inflation payouts for accounts + a payout to fee pool, so maybe change that to inflationPayouts?